MMM-MonthlyCalendar.js

```diff
@@ -105,48 +105,58 @@ Module.register("MMM-MonthlyCalendar", {
-        e.startDate = new Date(+e.startDate);
-        e.endDate = new Date(+e.endDate);
+      // Step 1: Parse and filter incoming events
+      self.sourceEvents[sender.identifier] = payload
+        .map((e) => {
+          e.startDate = new Date(+e.startDate);
+          e.endDate = new Date(+e.endDate);
 
-        if (e.fullDayEvent) {
-          e.endDate = new Date(e.endDate.getTime() - 1000);
+          if (e.fullDayEvent) {
+            e.endDate = new Date(e.endDate.getTime() - 1000);
 
-          if (e.startDate > e.endDate) {
-            e.startDate = new Date(e.endDate.getFullYear(), e.endDate.getMonth(), e.endDate.getDate(), 1);
-          } else {
-            e.startDate = new Date(e.startDate.getTime() + 60 * 60 * 1000);
-          }
-        }
-        // If not full day event, could still be multiple days.
-        // https://github.com/kolbyjack/MMM-MonthlyCalendar/issues/42
-        else {
-          // If time is longer than 24 hours
-          // Could instead check if start and end times are on different days?
-          if (((e.endDate.getTime() - e.startDate.getTime()) / 1000) > 86400) {
-            e.multiDayEvent = true;
-            //console.log(e.title + "is longer" + e.multiDayEvent);
+            if (e.startDate > e.endDate) {
+              e.startDate = new Date(e.endDate.getFullYear(), e.endDate.getMonth(), e.endDate.getDate(), 1);
+            } else {
+              e.startDate = new Date(e.startDate.getTime() + 60 * 60 * 1000);
+            }
           }
-        }
 
-        return e;
-      }).filter(e => {
-        return !self.config.hideCalendars.includes(e.calendarName);
-      });
+          return e;
+        })
+        .filter((e) => !self.config.hideCalendars.includes(e.calendarName));
 
+      // Step 2: Schedule update
       if (self.updateTimer !== null) {
         clearTimeout(self.updateTimer);
         ++self.skippedUpdateCount;
       }
 
       self.updateTimer = setTimeout(() => {
-        var today = new Date().setHours(12, 0, 0, 0).valueOf();
+        const today = new Date().setHours(12, 0, 0, 0).valueOf();
 
-        self.events = Object.values(self.sourceEvents).reduce((acc, cur) => acc.concat(cur), [])
+        // Step 3: Combine and sort events
+        self.events = Object.values(self.sourceEvents)
+          .flat()
           .sort((a, b) => {
-            return a.startDate - b.startDate;
+            if (a.startDate != b.startDate) return a.startDate - b.startDate;
+            if (a.endDate != b.endDate) return a.endDate - b.endDate;
+            return a.title.localeCompare(b.title);
           });
 
+        // Step 4: Remove duplicates using a hash table
+        if (self.config.hideDuplicateEvents) {
+          const seenEvents = new Map(); // Hash table for deduplication
+          self.events = self.events.filter((event) => {
+            const key = `${event.title}|${event.startDate.valueOf()}|${event.endDate.valueOf()}|${event.calendarName}`;
```

Comment:
This looks great, thank you for this PR!  I don't believe `calendarName` should be part of the key here, one of the use cases for event deduplication is when the same event appears on multiple calendars.

Reply:
Thanks good catch

Reply:
I think you are right tho so I made the change